### PR TITLE
OPNET-603: Fail haproxy test when all kubeapi servers are down simultaneously

### PR DESCRIPTION
**Pull request by mkowalski**
Implements https://issues.redhat.com/browse/OPNET-603 (mirrored at https://redhat.atlassian.net/browse/OPNET-603).

The on-prem haproxy monitor test added in #29149 so far only flaked when haproxy detected individual kube-apiserver backends as down, which is expected over the course of installation or upgrade. As suggested in https://github.com/openshift/origin/pull/29149#discussion_r1806784345, what really matters is whether a single haproxy instance saw all kube-apiserver backends down at the same time, because that means the API was completely unreachable through the on-prem loadbalancer.

Changes:

- `findFullAPIOutageWindows()`: sweeps over the constructed `OnPremHaproxyDetectsDown` intervals per haproxy instance, counting how many distinct kube-apiserver backends are concurrently down; windows with at least 3 backends down (on-prem HA control plane size) are full API outages. A backend recovering the same second another goes down is not counted as overlap, and touching windows are merged so a same-second flap does not split one outage into two.
- `evaluateFullAPIOutages()`: new junit test `[Jira: Networking / On-Prem Host Networking] Haproxy must not detect all kubeapi servers down simultaneously`. The first outage window for every haproxy instance is tolerated, since when haproxy starts during the installation all kube-apiservers are expected to be down until they come up for the first time. Any subsequent window fails the test with the node, time range and duration in the output.
- The existing flaky reachability test is left unchanged; the new result is appended in `EvaluateTestsFromConstructedIntervals`.
- Unit tests covering the sweep semantics (threshold, per-instance isolation, boundary/merge edge cases) and the junit evaluation (install-only tolerance, post-install failure).

Verified with `go build`, `go vet` and `go test ./pkg/monitortests/network/onpremhaproxy/...` (all tests pass).

## Summary by CodeRabbit

* **New Features**
  * Enhanced on-prem HAProxy monitoring with full API outage detection: identifies sustained “full outage” windows when multiple kube-apiserver backends are down concurrently (using a defined concurrency threshold per HAProxy node).
  * Updated generated monitoring reports to append a dedicated JUnit result for full outages, while conditionally emitting the existing kube-API “down” success/failure output based on observed intervals.

* **Tests**
  * Added new test suites covering full outage window calculation and evaluation, including partial-overlap, simultaneous backend failures, boundary-touching intervals, and startup-window skipping behavior.

### CODE REVIEW

File: pkg/monitortests/network/onpremhaproxy/monitortest.go [hunk 1]
```diff
@@ -211,22 +212,159 @@ func (*operatorLogAnalyzer) EvaluateTestsFromConstructedIntervals(ctx context.Co
...
+	ret := []*junitapi.JUnitTestCase{}
 	if !somethingFailed {
-		return []*junitapi.JUnitTestCase{success}, nil
+		ret = append(ret, success)
+	} else {
+		failure := &junitapi.JUnitTestCase{
+			Name: testName,
+			FailureOutput: &junitapi.FailureOutput{
+				//Message: fmt.Sprint("something happened with haproxy"),
+				Output: "Haproxy detected some kubeapi-servers down. It's not necessarily an issue, it's expected over the course of installation. Go and check messages. Look at intervals in sippy to see a full graph of which haproxy instance detected which kubeapi-server as down. Plotted on a time axis, you will see if at any point in time all the kubeapi-servers were down. Only then, it is an issue.",
```
Comment (dgoodwin) [inline pkg/monitortests/network/onpremhaproxy/monitortest.go]:
```suggestion
				Output: "Haproxy detected some kube api-servers were down. This is a normal occurrence during install/upgrade and not necessarily an issue unless they all are. Use interval charts for a timeline to see when this occurred.",
```

File: pkg/monitortests/network/onpremhaproxy/monitortest.go [hunk 2]
```diff
@@ -211,22 +212,159 @@ func (*operatorLogAnalyzer) EvaluateTestsFromConstructedIntervals(ctx context.Co
...
+			},
+			SystemOut: strings.Join(testNameToFailures[testName], "\n"),
+			//SystemErr: fmt.Sprintf("syserr; found %d lines in the failure map", len(testNameToFailures[testName])),
```
Comment (dgoodwin) [inline pkg/monitortests/network/onpremhaproxy/monitortest.go]: Let's kill off the commented lines.

File: pkg/monitortests/network/onpremhaproxy/monitortest.go [hunk 3]
```diff
@@ -211,22 +212,159 @@ func (*operatorLogAnalyzer) EvaluateTestsFromConstructedIntervals(ctx context.Co
...
+	type sweepEvent struct {
+		at    time.Time
+		delta int
+	}
+
+	eventsPerNode := map[string][]sweepEvent{}
+	for _, interval := range downIntervals {
+		// The locator key has the form "<node running haproxy>___<kube-apiserver backend>".
+		pairKey := interval.Locator.Keys[monitorapi.LocatorOnPremKubeapiUnreachableFromHaproxyKey]
+		parts := strings.SplitN(pairKey, "___", 2)
```
Comment (dgoodwin) [inline pkg/monitortests/network/onpremhaproxy/monitortest.go:262]: Do we need two pack two keys into one? I would expect explicit keys to be used.

File: pkg/monitortests/network/onpremhaproxy/monitortest.go [hunk 4]
```diff
@@ -211,22 +212,159 @@ func (*operatorLogAnalyzer) EvaluateTestsFromConstructedIntervals(ctx context.Co
...
+	return ret
+}
+
+// evaluateFullAPIOutages produces a junit result failing whenever a single haproxy instance
+// reported all kube-apiserver backends down at the same time. The first occurrence for every
+// haproxy instance is tolerated: when haproxy starts during the installation, all kube-apiservers
+// are expected to be down until they come up for the first time. Any later occurrence means the
+// API was completely unreachable through the on-prem loadbalancer.
+func evaluateFullAPIOutages(downIntervals monitorapi.Intervals) []*junitapi.JUnitTestCase {
+	const testName = "[Jira: Networking / On-Prem Host Networking] Haproxy must not detect all kubeapi servers down simultaneously"
```
Comment (dgoodwin) [inline pkg/monitortests/network/onpremhaproxy/monitortest.go]:
```suggestion
	const testName = "[Jira: Networking / On-Prem Host Networking] Haproxy should not encounter all kube apiservers down simultaneously"
```

Comment (dgoodwin) [reply pkg/monitortests/network/onpremhaproxy/monitortest.go]: Just avoiding the test name implying it's haproxy's fault for noticing :)

File: pkg/monitortests/network/onpremhaproxy/monitortest.go (hunk 3 above)
Comment (mkowalski) [reply pkg/monitortests/network/onpremhaproxy/monitortest.go:262]: Not sure if I understand you correctly, but... This monitor test builds 3x3 matrix of "how node X sees all the other nodes in the cluster". So the locator is built as `master1__master1`, `master1__master2`, `master1__master3`. This is a hack because of type of the struct we are using (we cannot have something like `map[string][string]` but only `map[string]`). So we pack 2 keys into one to know who tested whom.

File: pkg/monitortests/network/onpremhaproxy/monitortest.go (hunk 2 above)
Comment (mkowalski) [reply pkg/monitortests/network/onpremhaproxy/monitortest.go]: Done

File: pkg/monitortests/network/onpremhaproxy/monitortest.go (hunk 3 above)
Comment (dgoodwin) [reply pkg/monitortests/network/onpremhaproxy/monitortest.go:262]: I would have expected one interval per haproxy server, with locator key Node and perhaps something like ToNode. Are all results from all haproxies packed into one single interval?

Comment (mkowalski) [reply pkg/monitortests/network/onpremhaproxy/monitortest.go:262]: Not the case, maybe if we implemented it today that would be the elegant design.

https://github.com/openshift/origin/pull/29149/changes#diff-b1d9d027b846ecec1da525e3545683cbb4f0c641661e31b7a992ab50159ed7bbR105

This is my line from 2y ago where I packed it all together and used the magic string as a key.
